### PR TITLE
gitserver: update repo sizes only for repos that changed

### PR DESCRIPTION
**Pull request by mrnugget**
@eseliger and I noticed that we always do a full insert of possibly 10000 repositories when updating the repo size, even if that size didn't change, only updating it `on conflict`.

Since `repo` and `gitserver_repos` are consistent now, thanks to the database trigger, we can use normal `UPDATE` calls to update the size of the repos in `gitserver_repos`.

So what this does is it reduces the writes-to-disk that the database does and it fixes the reporting of "these repos have been updated" by only reporting when a repo _has_ in fact been updated in the database.

I think it's also clearer what the code does, vs. the previous use of `WithInserterWithReturner` where the `returner*` arguments where nil and the whole thing was only used because of the `ON CONFLICT`.

## Test plan

- Manual testing by checking log output locally, modifying database values, seeing how gitserver behaves, etc.
- Unit tests (old and new)

### CODE REVIEW

File: internal/database/gitserver_repos.go [hunk 1]
```diff
@@ -603,38 +603,70 @@ WHERE gr.repo_size_bytes IS NULL
...
+			return 0, err
+		}
+		updatedRows += int(rowsAffected)
 	}
-	return nil
+	return updatedRows, nil
 }
 
+const updateRepoSizesQueryFmtstr = `
+UPDATE gitserver_repos AS gr
```
Comment (eseliger) [inline internal/database/gitserver_repos.go:657]: query locator docstring pls

File: internal/database/gitserver_repos.go [hunk 2]
```diff
@@ -603,38 +603,67 @@ WHERE gr.repo_size_bytes IS NULL
...
+	currentCount := 0
+	updatedRows := 0
+	for repo, size := range repos {
+		batch[currentCount] = sqlf.Sprintf("(%s::integer, %s::bigint)", repo, size)
+
+		currentCount += 1
+
+		if currentCount == batchSize || currentCount == left {
+			// IMPORTANT: we only take the elements of batch up to currentCount
+			q := sqlf.Sprintf(updateRepoSizesQueryFmtstr, sqlf.Join(batch[:currentCount], ","))
```
Comment (sashaostrikov) [inline internal/database/gitserver_repos.go:632]: I believe that this code will always take "first $currentCount" elements from `batch` which will be more than the `batchSize` if `len(repos) > batchSize` and result in postgres error?

I guess we should change `batch[:currentCount]` to `batch[updatedRows:currentCount]`?

Comment (sashaostrikov) [reply internal/database/gitserver_repos.go:632]: If it is so, it's worth adding a unit test for `updateRepoSizesWithBatchSize` function with `len(repos) > batchSize`

**Review by sashaostrikov — COMMENTED** [review]
That's a great change, thanks! 👍 

I might have spotted a bug, please take a look at my comment

File: internal/database/gitserver_repos.go (hunk 1 above)
Comment (sashaostrikov) [reply internal/database/gitserver_repos.go]: This might still be not resolved

File: internal/database/gitserver_repos.go (hunk 2 above)
Comment (mrnugget) [reply internal/database/gitserver_repos.go:632]: Yup, that's tested here:

https://github.com/sourcegraph/sourcegraph/blob/c790866c23788b574f8c16f07e94c98654dd98de/internal/database/gitserver_repos_test.go#L1073-L1089

It's `len(repos) > batchSize` and `>= batchSize` and `< batchSize`

Comment (mrnugget) [reply internal/database/gitserver_repos.go:632]: > I guess we should change batch[:currentCount] to batch[updatedRows:currentCount]?

`currentCount` is always reset at the end of the loop once a batch has been updated.

Comment (sashaostrikov) [reply internal/database/gitserver_repos.go:632]: ah, I overlooked that batch array has its first `batchSize` elements overwritten 🤦

File: internal/database/gitserver_repos.go (hunk 1 above)
Comment (mrnugget) [reply internal/database/gitserver_repos.go]: It is: https://github.com/sourcegraph/sourcegraph/pull/39584/files/#diff-7e59d2cb7bca1c2f056a9dd00e374a89c41f8804a57c19f2601b46512ecb0cd8R653

Comment (sashaostrikov) [reply internal/database/gitserver_repos.go]: it's the end of the Thursday, it's just the end of the Thursday...

**Review by sashaostrikov — APPROVED** [review]
LGTM! Sorry for wrong comments 😬

File: internal/database/gitserver_repos.go [hunk 3]
```diff
@@ -603,38 +603,67 @@ WHERE gr.repo_size_bytes IS NULL
...
-	return nil
+
+	return updatedRows, nil
 }
 
+const updateRepoSizesQueryFmtstr = `
+-- source: internal/database/gitserver_repos.go:gitserverRepoStore.UpdateRepoSizes
+UPDATE gitserver_repos AS gr
+SET
+    repo_size_bytes = tmp.repo_size_bytes
```
Comment (mrnugget) [inline internal/database/gitserver_repos.go:656]: @mrnugget you need to update `updated_at`

**Pull request merged**